[addons] add OpenGL support to the addon interface and allow GL usage via DirectX

AlwinEsch opened the pull request:
## Description
This is a request to facilitate GL usage on the addon and to allow GL under DirectX.

It is not yet final and since the addon special CMake stuff is needed, it still needs to be clarified how they are provided.

This are required:
 - [FindOpenGLES.cmake](https://github.com/xbmc/screensaver.matrixtrails/blob/94a572ec87830604a736ceb2e89907589fa36034/FindOpenGLES.cmake#L18)
- [./depends/windows/angle/*](https://github.com/xbmc/screensaver.matrixtrails/blob/94a572ec87830604a736ceb2e89907589fa36034/depends/windows/angle)

To test:
https://github.com/AlwinEsch/screensaver.example.windows.gles

My ideas:
1. In "./xbmc/addons/kodi-addon-dev-kit" add a "docs (?)" Folder and lay the necessary stuff there
    - ***PRO*** In devkit folder everything deposited together and not distributed between different places.
    - ***PRO*** One understands better why a `HAS_ANGLE` is in the addon headern
    - ***CONTRA*** The code is not used in the construction therefore errors can creep in.
2. Create a Hello World (Triangle) demo addon which is referenced
    - ***PRO*** Can be used to test the headers
    - ***PRO*** Would be a working example for external developers
    - ***CONTRA*** There is something in Kodi which needs CMake stuff from outside.

What ideas do you have or what would you prefer?

---------------------------------------
The main part is the GL.h, to have the integration of the necessary header
OS independently of the addon. Depending on the OS and CMake definition,
this will provide the necessary includes.

With "GLonDX.h" a background class is added which allows to use GL also
under Windows with DirectX.
This is meant for the case that an addon can not easily be converted to
DirectX. If CMake defines a HAS_ANGLE this will be used. In addition,
however, a special "depend" is needed and a customized "FindOpenGLES.cmake".
The support is included in screensaver, visualization and CRendering.

In addition, this adds helper classes to the addon interface to facilitate
OpenGL usage "Shader.h"

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

Why I would like that so at the addon headern is to have it independent at the addon OS and keep `# ifdef` away from the addon. So an addon would be usable everywhere without any changes.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

afedchin reviewed (changes requested):
may be we would add some stub of `CGLonDX` in case if `HAS_ANGLE` undefined to avoid a lot of `#ifdef HAS_ANGLE`

AlwinEsch commented:
What you mean exactly with "stub"?

I'm not sure how best to do that without the "#ifdef".

I want to completely avoid "#ifdef" and code differences (for win) at the addon. To do this I had to integrate the calls into the "C" part which makes it harder. Wanted to make it about the classes at the beginning but so the addon would have to choose the right one for his case.

afedchin commented:
with stub I meant similar class but with empty methods. we can declare an interface and inherit from it the implementation and the stub.
something like the following
```cpp
struct IGLonDX
{
  // i-face definition
  virtual ~IGLonDX() = default;
  virtual bool InitGL(void*) = 0;
  ...
};
#ifdef HAS_ANGLE
// real implementation
class CGLonDXAngle : public IGLonDX
{
  bool InitGL(void* pDevice) override
  {
    ...
  }
  ...
};
using CGLonDX = CGLonDXAngle;
//#elif defined(TARGET_WINDOWS)
// some specific implementation 
#else
// class with empty methods
class CGLonDXStub : public IGLonDX
{
  bool InitGL(void*) override { return true; }
  ...
};
using CGLonDX = CGLonDXStub;
#endif
```

there will be only one `ifdef` in `GLonDX.h` file

AlwinEsch commented:
Is changed to your suggestion and now looks very much cleaner. Thanks a lot!

It also makes it easier to bring in other help systems (if any).

afedchin commented:
looks much better, but time changes should go to separate PR, no need mix stuff.

afedchin commented on xbmc/addons/ScreenSaver.cpp:
```diff
@@ -26,13 +23,9 @@ CScreenSaver::CScreenSaver(BinaryAddonBasePtr addonBase)
-#endif
   m_struct.props.x = 0;
   m_struct.props.y = 0;
+  m_struct.props.device = CServiceBroker::GetWinSystem()->GetHWContext();
```
I see no reason to keep this, an addon can use `GetHWContex()` now. the same for visualization.

AlwinEsch replied:
For me better to leave currently as backward compatibility for not changed addons.

afedchin commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/renderHelper.h:
```diff
@@ -0,0 +1,75 @@
+  if (CAddonBase::m_interface->addonBase->m_renderHelper)
+    return CAddonBase::m_interface->addonBase->m_renderHelper;
+
+  std::shared_ptr<kodi::gui::IRenderHelper> renderHelper = std::make_shared<CRenderHelper>();
```
iirc not all compilers support `std::make_shared` so use generic approach
```suggestion
  std::shared_ptr<kodi::gui::IRenderHelper> renderHelper(new CRenderHelper());
```

ksooo replied:
> iirc not all compilers support std::make_shared 

 Oh really? I use it all over the place in Kodi and never ran into problems.

afedchin replied:
I don't know exactly. may be I mixed up with `std::make_unique`

AlwinEsch commented:
Here the planned example addon about the use of this:
https://github.com/AlwinEsch/screensaver.example.windows.gles
If this request OK then the addon can add to xbmc on git as example.

afedchin commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/GL.h:
```diff
@@ -0,0 +1,105 @@
+#elif HAS_GLES >= 2
+  #define GL_TYPE_STRING "GLES"
+  #if defined(WIN32)
+    #error Use of GLES under Windows is not possible
```
what was changed?

AlwinEsch replied:
Oh shit, sorry. Was go away during update of docs, c&p in the night is not always not good, better to go to sleep ;-) and to make it in if you are rested.

Becomes fixed. I add currently the GL.h header to the addons itself, to have them prepared after is in Kodi itself (with only removing of them from addons)

AlwinEsch commented:
@afedchin the depend request is in and this is rebased and cleaned a bit.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/CMakeLists.txt:
```diff
@@ -1,7 +1,8 @@
             Window.h
-            definitions.h)
+            definitions.h
+            renderHelper.h)
```
I think it's quite unusual (if not violating our coding guidelines) to introduce source files starting with a lowercase letter.

AlwinEsch replied:
Here's a little special case for me, files that start with Uppercase are meant to be used directly from the addon. Those with lowercase are only headers that are used in the background by other headers, so first insignificant for addon developers.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/renderHelper.h:
```diff
@@ -0,0 +1,75 @@
+  if (CAddonBase::m_interface->addonBase->m_renderHelper)
+    return CAddonBase::m_interface->addonBase->m_renderHelper;
+
+  std::shared_ptr<kodi::gui::IRenderHelper> renderHelper(new CRenderHelper());
```
const std::shared_ptr

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/Shader.h:
```diff
@@ -0,0 +1,355 @@
+    }
+
+    // create program object
+    if (!(m_shaderProgram = glCreateProgram()))
```
I know it's personal style, but I hate assignments in "if" statements. This decreases readability and one has to take a deeper look to understand whether actually `==` was intended.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/Shader.h:
```diff
@@ -0,0 +1,355 @@
+
+    Free();
+
+    if (m_source.length()==0)
```
blank before and after '0'

AlwinEsch replied:
`if` becomes removed was a wrong leftover from old code.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/Shader.h:
```diff
@@ -0,0 +1,355 @@
+    return true;
+  }
+
+  bool OK() { return m_compiled; }
```
could the member function be 'const'?

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/GLonDX.h:
```diff
@@ -0,0 +1,366 @@
+    AddressV = CLAMP;
+    Comparison = NEVER;
+  };
+  float4 main(in float2 tex : TEXCOORD0) : SV_TARGET
```
empty line before function definition

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/GLonDX.h:
```diff
@@ -0,0 +1,366 @@
+    }
+
+    if (FAILED(pOffScreenTexture.As(&dxgiResource)) ||
+      FAILED(dxgiResource->GetSharedHandle(&sharedHandle)))
```
wrong indentation

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/GLonDX.h:
```diff
@@ -0,0 +1,366 @@
+    if (!createD3DResources())
+      return false;
+
+    if(eglMakeCurrent(m_eglDisplay, m_eglBuffer, m_eglBuffer, m_eglContext) != EGL_TRUE)
```
blank between if and (

AlwinEsch committed:
[addons] add OpenGL shader support

This implement a helper class to handle GL and GLES related loading
of Vertex and Fragment shaders.

The user can define files or inside code defined text or both together.

This commit depend with use of GL.h to the next commit

AlwinEsch commented:
Thanks @ksooo for review, your note places (and few others) are changed and code is cleaned up. During test it has worked :-)

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/GLonDX.h:
```diff
@@ -0,0 +1,369 @@
+/*
```
replace the windows line ending with an unix line ending in this line

AlwinEsch replied:
Such a crap, still one there :smile:

AlwinEsch replied:
A question to prevent this CRLF problem, makes it sense to create a request where a `.gitattributes` with `text eol=lf` becomes added to Kodi?

https://help.github.com/en/articles/dealing-with-line-endings

AlwinEsch committed:
[addons] add OpenGL support on addon interface / allow GL via DirectX

The main part is the GL.h, to have the integration of the necessary header
OS independently of the addon. Depending on the OS and CMake definition,
this will provide the necessary includes.

With "GLonDX.h" a background class is added which allows to use GL also
under Windows with DirectX.
This is meant for the case that an addon can not easily be converted to
DirectX. If CMake defines a HAS_ANGLE this will be used. In addition,
however, a special "depend" is needed and a customized "FindOpenGLES.cmake".
The support is included in screensaver, visualization and CRendering.
